src/KDMqtt/mqtt.cpp

```diff
@@ -0,0 +1,523 @@
+/*
+  This file is part of KDUtils.
+
+  SPDX-FileCopyrightText: 2024 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+  Author: Marco Thaller <marco.thaller@kdab.com>
+
+  SPDX-License-Identifier: MIT
+
+  Contact KDAB at <info@kdab.com> for commercial licensing options.
+*/
+#include "mqtt.h"
+#include <spdlog/spdlog.h>
+
+namespace KDMqtt {
+
+constexpr std::chrono::milliseconds c_miscTaskInterval = std::chrono::milliseconds(1000);
+
+using namespace KDFoundation;
+
+MqttLib::MqttLib()
+    : m_isInitialized{ false }
+    , m_mosquittoLib(&MosquittoLib::instance())
+{
+}
+
+MqttLib &MqttLib::instance()
+{
+    static MqttLib s_instance;
+    return s_instance;
+}
+
+int MqttLib::init()
+{
+    int result = MOSQ_ERR_UNKNOWN;
+
+    if (!m_isInitialized) {
+        result = m_mosquittoLib->init();
+        const auto hasError = checkMosquittoResultAndDoDebugPrints(result, "MqttLib::init()");
```

Comment:
This function is quite verbose and requires us to re-state the function name. I think we could replace it with a macro that would automatically handle source location and honour SPDLOG compile-time log enabling/disabling:
```cpp
#if SPDLOG_ACTIVE_LEVEL <= SPDLOG_LEVEL_ERROR
#define LOG_MOSQUITTO_ERROR(logger, file, line, func, error_string) \
    (logger)->log(spdlog::source_loc{ file, line, func }, spdlog::level::err, error_string)
#else
#define LOG_MOSQUITTO_ERROR (void)0
#endif

static bool checkMosquittoError(std::shared_ptr<spdlog::logger> logger, int ec, const char *file, int line, const char *func)
{
    const bool isError = (ec != MOSQ_ERR_SUCCESS);
    if (isError) {
        LOG_MOSQUITTO_ERROR(logger, file, line, func, mosquitto_strerror(ec));
    }
    return isError;
}
#define CHECK_AND_LOG_MOSQUITTO_RESULT(ec) checkMosquittoError(m_logger, ec, __FILE__, __LINE__, __FUNCTION__)
```

Then this callsite becomes:
```cpp
const auto hasError = CHECK_AND_LOG_MOSQUITTO_RESULT(result);
```

Reply:
Once we go to c++20 the macros of course will be gone as we'll be able to replace this with `std::source_location`